### PR TITLE
Add Unicode formatting tests

### LINKED ISSUE #11554: Write test to exclude UTF-8 BOM from files in tree
Write a test to exclude regressions from #11546, 0aaf74b

### PR DESCRIPTION
**Pull request by xtqqczze**
# PR Summary

* exclude UTF-8 BOM from text files
* move `test\common` to `test\formatting`

## PR Context

* Fix #11554
* exclude regressions from #11546

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by xtqqczze** [comment]
Tests failure due to UTF-8 BOM in `./test/common/utf8/utf8.tests.ps1`, will push a commit to remove after review.

File: test/static/utf8/utf8.tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        $bytes = [byte[]]::new(3)
+        $stream = [System.IO.File]::OpenRead($_)
+        $stream.Read($bytes, 0, 3) > $null
+        $stream.Close()
+        return $bytes[0] -eq 0xef -and $bytes[1] -eq 0xbb -and $bytes[2] -eq 0xbf
+    }
+
+    $nonBinaryFiles.ForEach{
+        Context "'$_'" {
+            It 'file has no BOM' {
```
Comment (iSazonov) [inline test/static/utf8/utf8.tests.ps1:24]: Why do we need Context?

Comment (xtqqczze) [reply test/static/utf8/utf8.tests.ps1:24]: we could add additional tests, e.g. detect unicode substition of space with nbsp (prevent regression from #12576)

Comment (iSazonov) [reply test/static/utf8/utf8.tests.ps1:24]: I see your point to group tests by file.
```suggestion
        Context "File: $_" {
            It 'No BOM' {
```

Comment (xtqqczze) [reply test/static/utf8/utf8.tests.ps1:24]: I copied the structure from `markdown-link.tests.ps1` but I'm not so sure if we need a seperate test for each file. There all 2368 files and there is an overhead.

**Comment by xtqqczze** [comment]
@iSazonov I took measurements from `PowerShell-CI-static-analysis`

e754367 takes 48.5 seconds to run
3c864ac takes 351 milliseconds

there is 138x overhead to run seperate tests

File: test/static/utf8/utf8.tests.ps1 (hunk 1 above)
Comment (iSazonov) [reply test/static/utf8/utf8.tests.ps1:24]: We could report only errors.

**Comment by iSazonov** [comment]
> there is over 100x overhead to run seperate tests...

Perster `Context` and `It` blocks slow down execution. I think we could report only errors.

**Comment by iSazonov** [comment]
@xtqqczze Please continue the PR.

**Comment by xtqqczze** [comment]
@iSazonov what is your opinion on moving `test/common` to `test/static`

**Comment by xtqqczze** [comment]
rebased to clean up commit history

**Comment by iSazonov** [comment]
> what is your opinion on moving test/common to test/formatting

Intention was to put in the folder all common tests, not only formatting/markdown.

**Comment by xtqqczze** [comment]
> 
> 
> > what is your opinion on moving test/common to test/formatting
> 
> Intention was to put in the folder all common tests, not only formatting/markdown.

I'm not quite sure what is "common" about such tests since these tests do not run in each pipeline, only in `PowerShell-CI-static-analysis`.

**Comment by iSazonov** [comment]
I hope @TravisEz13 makes a conclusion about "move test\common to test\formatting".

**Comment by TravisEz13** [comment]
The test should be under common.  We shouldn't create a new structure.  If a formatting folder is needed, make it under common.

**Comment by xtqqczze** [comment]
@TravisEz13 push forced to remove change to folder structure

File: test/common/FileEncoding/FileEncoding.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    It 'No Unicode BOM' {
+        filter hasBom {
+            $_.Where{
+                $bytes = [byte[]]::new(3)
+                $stream = [System.IO.File]::OpenRead($_)
+                $stream.Read($bytes, 0, 3) > $null
+                $stream.Close()
+                $bytes[0] -eq 0xef -and $bytes[1] -eq 0xbb -and $bytes[2] -eq 0xbf
```
Comment (iSazonov) [inline test/common/FileEncoding/FileEncoding.Tests.ps1:21]: There are some BOM kinds. I'd expect we check all.

Comment (xtqqczze) [reply test/common/FileEncoding/FileEncoding.Tests.ps1:21]: There is only one kind of BOM for UTF-8. Perhaps in new PR we could verify all Unicode files are in UTF-8, not UTF-16, UTF-32, etc.

Comment (iSazonov) [reply test/common/FileEncoding/FileEncoding.Tests.ps1:21]: I don't understand why does not check the 3 bytes for BOMs in one place.

Comment (iSazonov) [reply test/common/FileEncoding/FileEncoding.Tests.ps1]: The test should be "No BOM" and check a preambula for all cases.
https://www.powershellgallery.com/packages/PSTemplatizer/1.0.20/Content/Functions%5CGet-FileEncoding.ps1

**Comment by iSazonov** [comment]
@xtqqczze Please rebase to get latest CI updates.

File: test/common/FileEncoding/FileEncoding.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    AfterAll {
+        Pop-Location
+    }
+
+    It 'No UTF-8 BOM' {
+        filter hasUtf8Bom {
+            $_.Where{
+                $bom = [Text.UnicodeEncoding]::UTF8.GetPreamble()
```
Comment (iSazonov) [inline test/common/FileEncoding/FileEncoding.Tests.ps1:19]: I believe it is better to detect the encoding. C# code:
```c#
            Encoding GetPathEncoding()
            {
                using (StreamReader reader = new StreamReader(this.Path, Utils.utf8NoBom, detectEncodingFromByteOrderMarks: true))
                {
                    _ = reader.Read();
                    return reader.CurrentEncoding;
                }
            }
```

File: test/common/FileEncoding/FileEncoding.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (TravisEz13) [inline test/common/FileEncoding/FileEncoding.Tests.ps1:1]: We should actually run this test somewhere

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: test/common/FileEncoding/FileEncoding.Tests.ps1 (hunk 4 above)
Comment (TravisEz13) [reply test/common/FileEncoding/FileEncoding.Tests.ps1:1]: Perhaps, add another job to here
https://github.com/PowerShell/PowerShell/blob/master/.vsts-ci/misc-analysis.yml

**Or** add another workflow:

[workflows](https://github.com/PowerShell/PowerShell/tree/master/.github/workflows)

Whichever you are more comfortable with

**Pull request closed without merging**